smarts/core/sumo_road_network.py

```diff
@@ -946,28 +955,41 @@ def _snap_external_holes(self, lane_to_poly, snap_threshold=2):
             if lane.getEdge().isSpecial():
                 continue
 
-            incoming = self._graph.getLane(lane_id).getIncoming()
+            incoming = lane.getIncoming()
             if incoming and incoming[0].getEdge().isSpecial():
                 continue
 
-            outgoing = self._graph.getLane(lane_id).getOutgoing()
+            outgoing = lane.getOutgoing()
             if outgoing:
                 outgoing_lane = outgoing[0].getToLane()
                 if outgoing_lane.getEdge().isSpecial():
                     continue
 
             lane_shape = lane_to_poly[lane_id]
+            new_coords = []
             for x, y in lane_shape.exterior.coords:
                 for nl, dist in self.nearest_lanes(
                     Point(x, y),
                     include_junctions=False,
                 ):
-                    if (not nl) or (nl and nl.in_junction):
+                    if not nl or nl.in_junction or nl.lane_id == lane_id:
+                        continue
+                    ll, _ = nl.lane_to_left
+                    rl, _ = nl.lane_to_right
+                    if (not ll or ll.lane_id != lane_id) and (
+                        not rl or rl.lane_id != lane_id
+                    ):
                         continue
                     nl_shape = lane_to_poly.get(nl.lane_id)
                     if nl_shape:
-                        lane_shape = Polygon(snap(lane_shape, nl_shape, snap_threshold))
-            lane_to_poly[lane_id] = lane_shape
+                        p, np = nearest_points(shPoint(x, y), nl_shape)
+                        if p.distance(np) < snap_threshold:
+                            new_coords.append(np)
+                            break
```

Comment:
Is this to express snapping to an edge instead of points? I think we should filter out duplicate points here if the snap threshold repeats points in the polygon. I think this also still might run into a similar problem since we may have some points fall out of the `snap_threshold` that should still be used.

I think one likely problem that we have run into is that with curved roads the outer points of a internal lane might not line up with inner points of the outer lane. 

I think one option we have if the edges are parallel is to construct the outermost lane `lane0` in entirety and then reuse the inner edge of the previous lane `lane(n-1)` on each pass as we move towards the innermost lane `lane(n)`.